### ISSUE TITLE
re-train for higher resolution

### ISSUE THREAD

**Issue by alessiapacca** [AliaksandrSiarohin/first-order-model#298]
hello, I have a question on higher resolution usage of the model.

Is there the necessity to re-train the model if I want to go to higher resolution, or I can just change the config file and test with higher resolution inputs?

Moreover, how long did the training take for the checkpoints you uploaded? Did you train it on 4 gpus?

Thanks!

**Comment by AliaksandrSiarohin** [comment]
You can check the issue #20, training was done on 2 gpus for 1 or 2 days.

**Comment by alessiapacca** [comment]
Thanks for your answer! And if I want to train it for example on 8 gpus, I should only change the command to:

**CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 python run.py --config config/dataset_name.yaml --device_ids 0,1,2,3,4,5,6,7** ?

**Comment by AliaksandrSiarohin** [comment]
Yes, you should also change the batch_size in the config, usually larger batch lead to better performance.

**Comment by alessiapacca** [comment]
thank you. I have another question on the param **checkpoint_freq** of the config, what is the unit of measurement of it? is it measured in **num_epochs**, or **num_epochs*num_repeats**?

**Comment by AliaksandrSiarohin** [comment]
Epochs.

**Comment by ster-arch** [comment]
I only have 1 GPU with 4GB memory. To train 256x256 for my own mp4 training set, I could only use batch_size 1. Even batch_size 2 is causing out of memory. Is there a way to trade-off speed vs memory requirement? I would like to train for higher resolution (e.g. 512x512, with higher batch_size if possible), doesn't really matter if it takes longer. 

Is it possible? Which config parameter could be adjusted to reduce memory requirement of the GPU? Thanks

**Comment by AliaksandrSiarohin** [comment]
There is no easy way.

**Comment by alessiapacca** [comment]
@AliaksandrSiarohin Are you sure it only took 1 or 2 days to train on the entire vox dataset with 2 gpus, 100 epochs and 75 num_repeats? 
I am using 8 gpus 2080 and it is much slower than that. It took me 24 hours to train with 100 epochs and 10 num_repeats

**Comment by AliaksandrSiarohin** [comment]
Which resolution is that 256x256?

**Comment by alessiapacca** [comment]
Yes. 
resolution 256x256
batch size 72
num_repeats 10
100 epochs
Took 24 hours.

Now I am trying to use num_repeats 65 by starting from that checkpoint, but it is extremely low, I think it will not take less than 2 more days

**Comment by AliaksandrSiarohin** [comment]
There was a bug in file naming. Can you send me example of how the data was named in your dataset?

**Comment by alessiapacca** [comment]
first file is: id10001#1zcIwhmdeo4#000.mp4
second: id10001#7w0IBEWc9Qw#000.mp4
and so on...
last: id11248#yiNkInm9OKQ#002.mp4

**Comment by AliaksandrSiarohin** [comment]
Hmm, looks correct. What is gpu utilisation?

**Comment by alessiapacca** [comment]
![Screenshot 2020-10-05 at 10 07 10](https://user-images.githubusercontent.com/48483036/95054808-b98fe680-06f2-11eb-8331-83c392abdd09.png)


This is with batch size = 40, num_repeats = 65 and 100 epochs
I tried with batch size = 72 and it was crashing, so I had to reduce it

**Comment by AliaksandrSiarohin** [comment]
So utilisation is 0? Or it is just beginning of training?

**Comment by alessiapacca** [comment]
yes, at the end I realize I was not using the gpus correctly, thanks for your quick responses!!

I have one more question about the synthesis of a smile. For example, using the vox-dataset weights, if I test on an image of a face that is not smiling, and a driving video which is smiling, then the output pretty much fails to generate the smile of the target person (and viceversa: a smiling target person and a serious driving video). Did this also happen to you? Is this dependent on the found keypoints, or on the training dataset? 

I tried with different resolutions and it doesn't change the result



For example, in your gifs (in the README of the repo), you animated game of thrones characters and their teeth are there. I don't understand how that is possible, because I have tried with several target people at test time and it does not synthesize teeth or closed-mouths in a good way

**Comment by AliaksandrSiarohin** [comment]
Sure this is how the method in ```--relative``` mode works. It requires some initial alignment in pose.

**Comment by alessiapacca** [comment]
and this is ok, but all the gifs of the characters from Game of Thrones are generated from images with closed mouths, while the network was still able to synthesize the mouth opening and the teeth. Did you not use  **`--relative`** in that case? 


or maybe you trained for more than just 100 epochs and 75 num_repeats? I am using your pre-trained weights so I really don't see how  


I even tried with one of them, by running
```
python demo.py  --config config/vox-256.yaml --driving_video nameofmyvideo.mp4 --source_image nameofmyimage.png --checkpoint checkpoints/vox-cpk.pth.tar --relative --adapt_scale --find_best_frame
```

and this is the result I obtain, with Trump motion as the driving video:


![got](https://user-images.githubusercontent.com/48483036/95195821-5d52c280-07d7-11eb-88e1-8714983979e3.gif)

**Comment by AliaksandrSiarohin** [comment]
This is because of ```--find_best_frame```, remove it and you will get my result.

**Comment by alessiapacca** [comment]
Thank you very much @AliaksandrSiarohin for being super helpful with the answers.

I only have one last question: if I use a high resolution video and a high resolution image as input for a test, they of course get resized to 256x256 and the output has that resolution. But the output is pretty noisy. Is there a specific reason for the output to be that noisy? I mean, does it depend on the generator? Is it just a limitation of this method? 
Cause neither the video or the pictures are noisy at the start, but the output is

**Comment by AliaksandrSiarohin** [comment]
I don't understand what do you mean by noisy.

**Comment by alessiapacca** [comment]
What I mean, is that if I start from this quality low resolution picture

![trump](https://user-images.githubusercontent.com/48483036/95559986-e1e05380-0a18-11eb-9966-0ef8a804c27b.png)

and I take a quality HD video, that however I crop (and it loses quality while being cropped) 
![ellen1](https://user-images.githubusercontent.com/48483036/95561293-cd9d5600-0a1a-11eb-8ebf-275de75e9d3d.gif)


I get this output, which is way more blurred and noisier compared to the original image
![result1](https://user-images.githubusercontent.com/48483036/95561308-d2620a00-0a1a-11eb-83b6-7cf265a12175.gif)

Does this depend on the noise of the cropped driving video? it's really difficult to find a video that, after being cropped at 256x256, mantains is quality

**Comment by AliaksandrSiarohin** [comment]
This is not depend on the driving video, only movement depend on driving. It is blurrier, because reconstruction is not perfect.

**Comment by pzyss** [comment]
@alessiapacca hello，I also encountered the problem of GPU utilization is 0. How did you solve this?  Thanks.

**Comment by alessiapacca** [comment]
@pzyss Unfortunately, I saw that basically sometimes the utilisation is high and sometimes is 0. So in my case, if takes a very long time to train

**Comment by pzyss** [comment]
@alessiapacca  Thanks for your answer!  There is another question, does the retrained model had good performance with higher resolution like 512x512?  It seems has some weird warping artifacts with resolution 512x512,  could you show the results of your retrained model.

**Comment by alessiapacca** [comment]
@pzyss still haven't tried!

**Comment by pzyss** [comment]
thanks.

**Comment by alessiapacca** [comment]
@AliaksandrSiarohin 

could you confirm/fill this info with what you used for training? 

Pytorch version: 1.0.0
Cuda version: 10.0
GPU driver version: ?
Python version: 3.6?
More information about what GPUs you are using: ?

thank you!

**Comment by Bovey0809** [comment]
> There was a bug in file naming. Can you send me example of how the data was named in your dataset?

What kind of bug is it?

**Comment by AliaksandrSiarohin** [comment]
Missing Id prefix, e.g id00001#

**Comment by AliaksandrSiarohin** [comment]
Missing Id prefix, e.g id00001#
